### PR TITLE
fix: Mark as unread not working

### LINKED ISSUE #27274: User can't mark "Direct Messages" as unread
### Description:

When a user is marked as unread the user is marked as unread for a second if the chat is open.

### Steps to reproduce:

1. Open the chat with another user
Screenshot from 2022-11-16 08-30-06
2. Click on "Mark Unread"
Screenshot from 2022-11-16 08-32-16

### Expected behavior:
The user is marked as unread.

### Actual behavior:
For one second the user is marked as unread:
Screenshot from 2022-11-16 08-36-13
but after the second it looks like this:
Screenshot from 2022-11-16 08-37-59

### Server Setup Information:

- Version of Rocket.Chat Server: 
- Operating System: FreeBSD 12.3
- Deployment Method: docker
- Number of Running Instances: 1
- DB Replicaset Oplog: Enabled
- NodeJS Version: v14.19.3
- MongoDB Version: 5.0.13

### Client Setup Information

- Desktop Browser Version: Google Chrome (107.0.5304.110-1), Mozilla Firefox (106.0.5)
- Operating System: Ubuntu 22.04.1 LTS

### PR DESCRIPTION
**Pull request by MartinSchoeler**
## Proposed changes (including videos or screenshots)

- Fixed issue where when you marked a room as unread and you were part of it, sometimes it would mark it as read right after

## Issue(s)

Closes #27274

[SUP-567](https://rocketchat.atlassian.net/browse/SUP-567)

## Steps to test or reproduce

This seems to happen only on busier servers, you can reproduce it on open.rocket.chat if you have the room open and mark it as unread, soon after it is marked as read again

[SUP-567]: https://rocketchat.atlassian.net/browse/SUP-567?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/meteor/app/message-mark-as-unread/client/actionButton.ts
Comment (MarcosSpessatto) [inline apps/meteor/app/message-mark-as-unread/client/actionButton.ts:1]: Do we need to delete this in this PR? Would it be possible to be removed in another one?

File: apps/meteor/client/sidebar/RoomMenu.tsx [hunk 1]
```diff
@@ -171,19 +171,24 @@ const RoomMenu = ({
...
 			if (subscription == null) {
 				return;
 			}
+
 			LegacyRoomManager.close(subscription.t + subscription.name);
 
-			router.navigate('/home');
+			if (roomOpen) {
+				router.navigate('/home');
+			}
```
Comment (MarcosSpessatto) [inline apps/meteor/client/sidebar/RoomMenu.tsx:189]: Is this related to the original issue? If not, let's create a specific PR for that

File: apps/meteor/client/sidebarv2/RoomMenu.tsx [hunk 2]
```diff
@@ -177,13 +177,18 @@ const RoomMenu = ({
...
 			if (subscription == null) {
 				return;
 			}
+
 			LegacyRoomManager.close(subscription.t + subscription.name);
 
-			router.navigate('/home');
+			if (roomOpen) {
+				router.navigate('/home');
+			}
```
Comment (MarcosSpessatto) [inline apps/meteor/client/sidebarv2/RoomMenu.tsx:189]: Same

File: apps/meteor/tests/e2e/channels-sidebar.spec.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+		await expect(sidebarItem.locator('.rcx-badge')).not.toBeVisible();
+	});
+
+	test('should mark a populated room as unread', async () => {
+		// focus should be on the next item
+		await poHomeChannel.sidenav.openChat(targetChannel);
+		await poHomeChannel.content.sendMessage('this is a message for reply');
+		const sidebarItem = await poHomeChannel.sidenav.selectMarkAsUnread(targetChannel);
+		await expect(sidebarItem.locator('.rcx-badge')).toBeVisible();
+	});
```
Comment (MarcosSpessatto) [inline apps/meteor/tests/e2e/channels-sidebar.spec.ts:37]: Can we add a test case using the mark as unread from the message action as well?

**Review by MarcosSpessatto — CHANGES_REQUESTED** [review]

File: apps/meteor/tests/e2e/page-objects/fragments/home-sidenav.ts [hunk 4]
```diff
@@ -68,6 +68,14 @@ export class HomeSidenav {
 		return this.page.locator(`[data-qa="sidebar-item"][aria-label="${name}"]`);
 	}
 
+	async selectMarkAsUnread(name: string) {
+		const sidebarItem = this.getSidebarItemByName(name);
+		await sidebarItem.focus();
+		await sidebarItem.locator('.rcx-sidebar-item__menu').click();
+		await this.page.getByRole('option', { name: 'Mark Unread' }).click();
+		return sidebarItem;
```
Comment (dougfabris) [inline apps/meteor/tests/e2e/page-objects/fragments/home-sidenav.ts:76]: I would suggest avoiding returning the sidebarItem here, and in order to get it use the `getSidebarItemByName` where you need

File: apps/meteor/tests/e2e/mark-unread.spec.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+			await poHomeChannelUser2.content.sendMessage('this is a message for reply');
+		});
+
+		test('should mark a populated room as unread', async () => {
+			await poHomeChannel.sidenav.openChat(targetChannel);
+
+			await poHomeChannel.content.openLastMessageMenu();
+			await poHomeChannel.page.locator('role=menuitem[name="Mark Unread"]').click();
+			const sidebarItem = poHomeChannel.sidenav.getSidebarItemByName(targetChannel);
+			await expect(sidebarItem.locator('.rcx-badge')).toBeVisible();
```
Comment (dougfabris) [inline apps/meteor/tests/e2e/mark-unread.spec.ts:55]: create the badge locator in page objects to re-use it across the tests

File: apps/meteor/tests/e2e/mark-unread.spec.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+			await poHomeChannelUser2.sidenav.openChat(targetChannel);
+			await poHomeChannelUser2.content.sendMessage('this is a message for reply');
+		});
+
+		test('should mark a populated room as unread', async () => {
+			await poHomeChannel.sidenav.openChat(targetChannel);
+
+			await poHomeChannel.content.openLastMessageMenu();
+			await poHomeChannel.page.locator('role=menuitem[name="Mark Unread"]').click();
```
Comment (dougfabris) [inline apps/meteor/tests/e2e/mark-unread.spec.ts:53]: create the markUnread locator in page objects to re-use it across the tests

**Review by MarcosSpessatto — DISMISSED** [review]

**Comment by dougfabris** [comment]
We should investigate more, mark unread feature through the message seems to be working partially. 
Sometimes the e2e is failing locally!

**Review by dougfabris — APPROVED** [review]

**Comment by ggazzo** [comment]
> Fixed issue where when you marked a room as read

UNread rigth?

**Review by dougfabris — DISMISSED** [review]

**Review by MarcosSpessatto — APPROVED** [review]

**Review by tiagoevanp — APPROVED** [review]

**Review by MarcosSpessatto — APPROVED** [review]

**Pull request merged**